lib/craft/clock_bound/client.ex

```diff
@@ -0,0 +1,233 @@
+  """
+
+  @default_shm_path "/var/run/clockbound/shm"
+  @clockbound_restart_grace_period 5 * 1_000_000_000
+
+  # 'AMZN CB'
+  @magic_0 0x414D5A4E
+  @magic_1 0x43420200
+
+  defstruct [
+    :earliest,
+    :latest,
+    :clock_status
+  ]
+
+  @type clock_status :: :unknown | :synchronized | :free_running | :invalid
+  @type t :: %__MODULE__{
+          earliest: DateTime.t(),
+          latest: DateTime.t(),
+          clock_status: clock_status()
+        }
+
+  @doc """
+  Returns the current clock time with error bounds and clock status.
+  """
+  def now(shm_path \\ @default_shm_path) do
+    with {:ok, data} <- read(shm_path),
+         real = :os.system_time(:nanosecond),
+         monotonic = monotonic_time(),
+         {:ok, {earliest, latest, clock_status}} <- compute_bound_at(real, monotonic, data),
+         {:ok, earliest} <- DateTime.from_unix(earliest, :nanosecond),
+         {:ok, latest} <- DateTime.from_unix(latest, :nanosecond) do
+      {:ok,
+       %__MODULE__{
+         earliest: earliest,
+         latest: latest,
+         clock_status: clock_status
+       }}
+    else
+      {:error, reason} -> {:error, reason}
+      _ -> {:error, :unknown}
+    end
+  end
+
+  @doc """
+  Compares two ClockBound windows and returns:
+
+    * `:lt` if a is strictly younger than b
+    * `:gt` if a is strictly older than b
+    * `:ov` if the ranges overlap and no strict ordering can be determined
+  """
+  def compare(cbe_1, cbe_2) do
+    cond do
+      DateTime.after?(cbe_1.earliest, cbe_2.latest) -> :gt
+      DateTime.before?(cbe_1.latest, cbe_2.earliest) -> :lt
+      true -> :ov
+    end
+  end
+
+  @doc """
+  Returns the current monotonic time in nanoseconds
+  For macOS, it uses the native NIF to get the monotonic time.
+  For other platforms, it uses the Erlang system_info function.
+  """
+  def monotonic_time do
+    case :os.type() do
+      {:unix, :darwin} ->
+        to_nanosecond(ClockBound.Native.get_monotonic_time())
+
+      _ ->
+        :erlang.system_info(:os_monotonic_time_source)[:time]
+    end
+  end
+
+  defp compute_bound_at(real, monotonic, %{
+         as_of: as_of,
+         void_after: void_after,
+         clock_status: status,
+         max_drift: max_drift,
+         bound: bound
+       }) do
+    # If max_drift is too large, it indicates that some issue in the clockbound data
+    if max_drift > 1_000_000_000 do
+      {:error, :max_drift_too_large}
+    else
+      as_of_ns = to_nanosecond(as_of)
+      void_after_ns = to_nanosecond(void_after)
+
+      # Validate whether the clockbound data read from shared memory is still valid at the time of the request by
+      # checking if the current monotonic time is lesser than the void_after time.
+      clock_status =
+        case status do
+          s when s in [:synchronized, :free_running] ->
+            cond do
+              # If the clockbound data has not been updated in a while, consider it trusted only upto a grace period of 5 sec.
```

Comment:
```suggestion
              # If the clockbound data has not been updated in a while, consider it trusted only up to a grace period of 5 sec.
```

is there a reason to offer any grace period at all?

Reply:
The grace period is to allow clockbound data be considered for a short grace period when the clockbound daemon dies and its restarting. This logic is taken from the original rust client. 

We can ignore this case and let it fail also? The only advantage i feel is the impact of this in our check. If the clockbound deamon restarts fast within the grace period than it will not disrupt our application flow.

Reply:
yeah, i think i'd rather let it fail instead. time estimation is already a messy business, it'd prefer to learn conservative with it.